src/lib/utils/meta_tables/meta_system_utilization_table.cpp

```diff
@@ -0,0 +1,353 @@
+
+  struct rusage resource_usage;
+  if (getrusage(RUSAGE_SELF, &resource_usage)) {
+    Fail("Unable to access rusage");
+  }
+
+  uint32_t nano = 1'000'000'000, micro = 1'000;
+  uint64_t system_time = resource_usage.ru_stime.tv_sec * nano + resource_usage.ru_stime.tv_usec * micro;
+  uint64_t user_time = resource_usage.ru_utime.tv_sec * nano + resource_usage.ru_utime.tv_usec * micro;
+
+  auto used = (user_time - last_user_time) + (system_time - last_system_time);
+  auto total = (clock_time - last_clock_time) * info.numer / info.denom;
+
+  last_clock_time = clock_time;
+  last_user_time = user_time;
+  last_system_time = system_time;
+
+  return (100.0f * used) / (total);
+
+#endif
+
+  Fail("Method not implemented for this platform");
+}
+
+MetaSystemUtilizationTable::SystemMemoryUsage MetaSystemUtilizationTable::_get_system_memory_usage() {
+#if defined __linux__
+
+  struct sysinfo memory_info;
+  sysinfo(&memory_info);
+
+  MetaSystemUtilizationTable::SystemMemoryUsage memory_usage;
+  memory_usage.total_ram = memory_info.totalram * memory_info.mem_unit;
+  memory_usage.total_swap = memory_info.totalswap * memory_info.mem_unit;
+  memory_usage.free_ram = memory_info.freeram * memory_info.mem_unit;
+  memory_usage.free_swap = memory_info.freeswap * memory_info.mem_unit;
+  memory_usage.total_memory = memory_usage.total_ram + memory_usage.total_swap;
+  memory_usage.free_memory = memory_usage.free_ram + memory_usage.free_swap;
+
+  return memory_usage;
+
+#elif defined __APPLE__
+
+  int64_t physical_memory;
+  size_t size = sizeof(physical_memory);
+  if (sysctlbyname("hw.memsize", &physical_memory, &size, nullptr, 0) != 0) {
+    Fail("Unable to call sysctl hw.memsize");
+  }
+
+  // Attention: total swap might change if more swap is needed
+  xsw_usage swap_usage;
+  size = sizeof(swap_usage);
+  if (sysctlbyname("vm.swapusage", &swap_usage, &size, nullptr, 0) != 0) {
+    Fail("Unable to call sysctl vm.swapusage");
+  }
+
+  vm_size_t page_size;
+  vm_statistics64_data_t vm_statistics;
+  mach_msg_type_number_t count = sizeof(vm_statistics) / sizeof(natural_t);
+
+  if (host_page_size(mach_host_self(), &page_size) != KERN_SUCCESS ||
+      host_statistics64(mach_host_self(), HOST_VM_INFO, reinterpret_cast<host_info64_t>(&vm_statistics), &count) !=
+          KERN_SUCCESS) {
+    Fail("Unable to access host_page_size or host_statistics64");
+  }
+
+  MetaSystemUtilizationTable::SystemMemoryUsage memory_usage;
+  memory_usage.total_ram = physical_memory;
+  memory_usage.total_swap = swap_usage.xsu_total;
+  memory_usage.free_swap = swap_usage.xsu_avail;
+  memory_usage.free_ram = vm_statistics.free_count * page_size;
+
+  // auto used = (vm_statistics.active_couunt + vm_statistice.inactive_count + vm_statistics.wire_count) * page_size;
+
+  return memory_usage;
+
+#endif
+
+  Fail("Method not implemented for this platform");
+}
+
+#if defined __linux__
+int64_t MetaSystemUtilizationTable::_int_from_string(std::string input_string) {

```

Comment:
Why is this necessary and why is it only implemented for Linux?

Reply:
this is for parsing lines in the /proc/self/status file
under macOS, we can get the values more easily form a system call

Reply:
Let's wait for the comment. You might want to reconsider the naming, too. Looks like it does more than just convert a string into an int.